File: src/main/java/aoc2019/Day14.java
package aoc2019;


import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import org.junit.jupiter.api.Assertions;


/*
 * Day 14: Space Stoichiometry
 * https://adventofcode.com/2019/day/14
 *
 */
public class Day14 {

	static final String FUEL = "FUEL";
	static final String ORE = "ORE";
	
	
	static class Product {
		int amount;
		String name;
		
		public Product(int amount, String name) {
			this.amount = amount;
			this.name = name;
		}

		static Product parse(String s) {
			var f = s.trim().split(" ");
			return new Product(Integer.parseInt(f[0]), f[1]);
		}
		
		@Override
		public String toString() {
			return String.format("%d %s", amount, name);
		}
		
		@Override
		public boolean equals(Object obj) {
			Product p = (Product)obj;
			return name.equals(p.name) && amount == p.amount;
		}
	}
	
	static class React {
		Product output;
		List<Product> input;
		long factor = 0;
		
		static React parse(String s) {
			var f = s.trim().split(" => ");
			React p = new React();
			p.input = Arrays.stream(f[0].split(",")).map(Product::parse).collect(Collectors.toList());
			p.output = Product.parse(f[1]);
			return p;
		}
				
		void setFactor(long targetAmount) {
			factor = targetAmount / output.amount;
			if (targetAmount % output.amount != 0) factor++; 
		}
		
		int amountNeeded(String chem) {
			return input.stream().filter(i -> i.name.equals(chem)).mapToInt(i -> i.amount).sum();
		}

		boolean needs(String chem) {
			return input.stream().anyMatch(i -> chem.equals(i.name));
		}

		@Override
		public String toString() {
			return String.format("%s => %s", input.stream().map(Product::toString).collect(Collectors.joining(", ")), output);
		}
	}
	
	public static void main(String[] args) throws Exception {
        System.out.println("=== tests 1 ===");
        Assertions.assertEquals(31, new Day14("input14a.txt").part1());
        Assertions.assertEquals(165, new Day14("input14b.txt").part1());
        Assertions.assertEquals(13312, new Day14("input14c.txt").part1());
        Assertions.assertEquals(180697, new Day14("input14d.txt").part1());
        Assertions.assertEquals(2210736, new Day14("input14e.txt").part1());
        
        System.out.println("=== tests 2 ===");
        Assertions.assertEquals(82892753, new Day14("input14c.txt").part2());
        Assertions.assertEquals(5586022, new Day14("input14d.txt").part2());
        Assertions.assertEquals(460664, new Day14("input14e.txt").part2());
        
        System.out.println("=== part1 ===");
        new Day14("input14.txt").part1();

        System.out.println("=== part2 ===");
        new Day14("input14.txt").part2();
	}
	

	List<React> reactions;
	Set<React> todo = new HashSet<>();

	Day14(String fname) throws Exception {
		reactions = Util.stringStreamOf(fname).map(React::parse).collect(Collectors.toList());
	}
	
	long part1() {	
		long ore = calcOre(1);
		System.out.printf("==> consumed %d ORE%n", ore);
		
		reactions.stream().filter(r -> List.of("FQPNW", "CQRJP", "GRQTX").contains(r.output.name))
		.forEach(r -> System.out.printf("%d  x  %s%n", r.factor, r.toString()));
		return ore;
	}

	
	int part2() {
		long teraOre = 1000000000000L;
		
		// binary search lower=0, upper=MAX
		int lb = 0;
		int ub = Integer.MAX_VALUE;
		int fuel = 0;
		long ore = 0;
		while (lb < ub-1) {
			fuel = lb + (ub-lb)/2;
			ore = calcOre(fuel);
			if (ore > teraOre) {
				ub = fuel;
			} else {
				lb = fuel;
			}
		}
		fuel = lb;
		System.out.printf("==> %d ORE produced %d FUEL%n", calcOre(fuel), fuel);
		return fuel;
	}

	// how much of chem is required as input by the current reactions
	long totalInputAmount(String chem) {
		return reactions.stream().mapToLong(r -> r.factor * r.amountNeeded(chem)).sum();
	}
	
	React findReact(String chem) {
		return reactions.stream().filter(r -> r.output.name.equals(chem)).findFirst().get();
	}

	long calcOre(long fuel) {
		todo.clear();		
		reactions.forEach(r -> r.factor = 0);
		
		var fuelReaction = findReact(FUEL);
		fuelReaction.factor = fuel;

		// check all other reactions
		todo.addAll(reactions);
		todo.remove(fuelReaction);
		
		// plan reactions
		while ( ! todo.isEmpty()) {
			for (React r : todo) {
				// required output of r is known - no TODO reactions require this product
				boolean cando = todo.stream().noneMatch(t -> t.needs(r.output.name));
				if (cando) {
					// how much of r output is needed for other reactions, how often must r repeat
					long need = totalInputAmount(r.output.name); 
					r.setFactor(need);
					
					todo.remove(r);
					break;
				}				
			}
		}
		
		return totalInputAmount(ORE);
	}

}

File: src/main/julia/day14.jl
Chem = AbstractString

struct Prod
    qty::Int
    name::Chem
end

struct Reac
    output::Prod
    input::Array{Prod}
end

function parseprod(x) 
    s = split(x, " ")
    Prod(parse(Int, s[1]), s[2])
end

function parsereact(x) 
    s = split(x, " => ")
    input = collect(parseprod(x) for x=split(s[1], ", "))
    output = parseprod(s[2])
    Reac(output, input)
end

# parseprod("2 GPBXP")
# parsereact("3 SMLPV, 11 NZJV, 1 HTSXK => 2 GPBXP")

R = Dict(r.output.name=>r for r=(parsereact(x) for x=readlines("input14.txt"))) # reaction map, eg. R["FUEL"]
N = Dict{Reac,Int}() # reaction factor for planned reactions

done(r::Reac) = r in keys(N)
todo(r::Reac) = !done(r)
done() = filter(done, collect(values(R)))
todo() = filter(todo, collect(values(R)))

# calc iterations of reaction to produce needed quantity
nreact(r::Reac, qty) = ceil(qty / r.output.qty) 

needed(p::Prod, chem::Chem) = p.name==chem ? p.qty : 0
needed(r::Reac, chem::Chem) = sum(needed(pin, chem) for pin=r.input)
# total amount of chem needed by all reactions
needed(c::Chem) = sum(N[r] * needed(r,c) for r=done())

needs(r::Reac, chem::Chem) = any(pin.name==chem for pin=r.input)
needs(r::Reac, pin::Prod) = needs(r, pin.name)

# no "to do" reaction depends on this reaction (needs the output)
ready(r::Reac) = any(needs(x, r.output) for x=done()) && !any(needs(x, r.output) for x=todo())
ready(c::Chem) = ready(R[c])

function produce(qty, result)
    global N = Dict(R[result]=>qty)

    # schedule reactions: as soon as we know the total required output
    while !isempty(todo())
        for r = filter(ready, todo())
            # how much of r's output do we need?
            need = needed(r.output.name)
            # how many times do we have to repeat the reaction?
            N[r] = nreact(r, need)
            # println("produce (", need, " ", r.output.name, ") <- ", N[r], " x ", r)
        end
    end

    return needed("ORE")
end

@time part1 = produce(1, "FUEL")


# part 2


# binary search for maximum parameter for predicate f
function binse(f::Function, (l,u)) 
    if (u <= l+1) return l; end
    m = (u+l) ÷ 2
    return f(m) ? binse(f, (m,u)) : binse(f, (l,m))
end
# binse(x->x<398, (0,1e9))

qore = 1000000000000
enoughOre(x) = produce(x, "FUEL") <= qore

@time part2 = binse(enoughOre, (0, qore))